src/lavinmq/mqtt/client.cr

```diff
@@ -0,0 +1,257 @@
+        @log = Logger.new(Log, @metadata)
+        @broker.vhost.add_connection(self)
+        @log.info { "Connection established for user=#{@user.name}" }
+        spawn read_loop
+      end
+
+      def client_name
+        "mqtt-client-#{@client_id}"
+      end
+
+      private def read_loop
+        socket = @socket
+        if socket.responds_to?(:"read_timeout=")
+          socket.read_timeout = (@keepalive * 1.5).seconds # 1.5 according to [MQTT-3.1.2-24].
+        end
+        loop do
+          @log.trace { "waiting for packet" }
+          packet = read_and_handle_packet
+          # The disconnect packet has been handled and the socket has been closed.
+          # If we dont breakt the loop here we'll get a IO/Error on next read.
+          if packet.is_a?(MQTT::Disconnect)
+            @log.debug { "Recieved disconnect" }
+            break
+          end
+        end
+      rescue ex : ::MQTT::Protocol::Error::PacketDecode
+        @log.warn(exception: ex) { "Packet decode error" }
+        publish_will
+      rescue ex : ::IO::TimeoutError
+        @log.warn { "Keepalive timeout (keepalive:#{@keepalive}): #{ex.message}" }
+        publish_will
+      rescue ex : ::IO::Error
+        @log.error { "Client unexpectedly closed connection: #{ex.message}" } unless @closed
+        publish_will
+      rescue ex
+        @log.error(exception: ex) { "Read Loop error" }
+        publish_will
+      ensure
+        @broker.remove_client(self)
+        @waitgroup.done
+        close_socket
+      end
+
+      def read_and_handle_packet
+        packet : MQTT::Packet = MQTT::Packet.from_io(@io)
+        @log.trace { "Recieved packet:  #{packet.inspect}" }
+        @recv_oct_count += packet.bytesize
+
+        case packet
+        when MQTT::Publish     then recieve_publish(packet)
+        when MQTT::PubAck      then recieve_puback(packet)
+        when MQTT::Subscribe   then recieve_subscribe(packet)
+        when MQTT::Unsubscribe then recieve_unsubscribe(packet)
+        when MQTT::PingReq     then receive_pingreq(packet)
+        when MQTT::Disconnect  then return packet
+        else                        raise "received unexpected packet: #{packet}"
+        end
+        packet
+      end
+
+      def send(packet)
+        @lock.synchronize do
+          packet.to_io(@io)
+          @socket.flush
+          @send_oct_count += packet.bytesize
+        end
+      end
+
+      def receive_pingreq(packet : MQTT::PingReq)
+        send MQTT::PingResp.new
+      end
+
+      def recieve_publish(packet : MQTT::Publish)
+        @broker.publish(packet)
+        # Ok to not send anything if qos = 0 (fire and forget)
+        if packet.qos > 0 && (packet_id = packet.packet_id)
+          send(MQTT::PubAck.new(packet_id))
+        end
+      end
+
+      def recieve_puback(packet : MQTT::PubAck)
+        @broker.sessions[@client_id].ack(packet)
```

Comment:
This looks odd. Why don't we do `@broker.puback(...)` instead of accessing `sessions`? We're not doing that for other packets.

Reply:
will fix next up